docs/contributing/backward-compatibility-promise.md

```diff
@@ -0,0 +1,144 @@
+
+### PHP Code
+Rules for PHP code are fully covered by [Symfony Backward Compatibility Promise](https://symfony.com/doc/3.4/contributing/code/bc.html).
+
+### Database Migrations
+A new version may include database migrations if the structure of [the entities](/docs/introduction/entities.md) changed.
+
+Migrations in `MINOR` versions are backward-compatible.
+It means they may not change types of existing columns, rename columns and tables, and remove nullability of a column.
+
+Migrations in `PATCH` versions are backward-compatible and may be used only for bug fixes.
+
+You should always check and test the database migrations before running them on your production data.
+
+*Tip: Read about the possibilities of altering the execution of DB migration using the [`migrations-lock.yml` file](/docs/introduction/database-migrations.md#locking-the-order-of-migrations).*
+
+### Translation Messages
+New [translation messages](/docs/introduction/translations.md) may be added or have its translation changed in any release.
+However, they may be removed or have their message ID changed only in `MAJOR` releases.
+
+When changing the message (eg. because of a typo) in a `MINOR` or `PATCH` release, only its translation may be changed.
+In such instances, it is preferred to keep using the original message ID to ensure backward compatibility with existing user-defined translation.
+This might be unintuitive for contributors because we use English text as message IDs.
+See an example of fixing a typo in an English translation message:
+
+Initially, there is only the message ID in the `messages.en.po` (translation doesn't have to be filled):
+```diff
+  msgid "Exaple translation"
+  msgstr ""
+```
+
+In a `PATCH` or `MINOR` version, the original message ID may not be removed:
+```diff
+  msgid "Exaple translation"
+- msgstr ""
++ msgstr "Example translation"
+```
+
+In a `MAJOR` version, the original message ID with the typo may be removed:
+```diff
+- msgid "Exaple translation"
+- msgstr "Example translation"
++ msgid "Example translation"
++ msgstr ""
+```
+
+### Routing
+New routes may be added in any release.
+Existing routes should not be modified, except for adding a new [optional placeholder](https://symfony.com/doc/3.4/routing/optional_placeholders.html).
+
+Route names may be changed only in a `MAJOR` release.
+
+### Docker Configuration and Orchestration Manifests
+Configuration of containers and orchestration is mostly in the `project-base` repository, which means that [it's excluded form the BC promise](#project-base-repository).
+
+The changes should always be described in [upgrade notes](/UPGRADE.md) (in the *Infrastructure* section).
+
+### Twig
+Twig blocks, functions, filters and the templates themselves can be removed or renamed only in a `MAJOR` release.
+
+Changes of Twig functions and filters in `MINOR` and `PATCH` releases must be backward-compatible.
```

Comment:
What about HTML structure? How should we deal with changing for example templates in admin (eg. add surrounding `<div>`) and changing blocks (eg. adding surrounding `<picture>` element)?

Reply:
Maybe there should be a new section ### HTML

Reply:
Not sure how to deal with that and couldn't find any applicable best practices in other projects.

I would prefer not dealing with it now and add new HTML-related rules when we have some experience. Maybe we find out that it's not even needed or not possible at all...